### PR TITLE
Docs: Adopting Dagger

### PR DESCRIPTION
**Pull request by shykes**
Expand the "Adopting Dagger" page with a framework for successful adoption.

### CODE REVIEW

File: docs/current_docs/adopting.mdx [hunk 1]
```diff
@@ -4,13 +4,121 @@ slug: /adopting
...
+Once you have [completed the quickstart](quickstart), and learned the basic concepts, it's time to take the next step: adopting Dagger in your project - what we call *daggerizing*. But how does one Daggerize a project, exactly?
 
-We get it, looking at that many lines of YAML can be intimidating. 
-Since each pipeline is unique in tooling and functionally, our general answer is to "start simple". Pick a part of your pipeline to Daggerize first, and then continue to Daggerize from there.
+Every project is different, and every pipeline is a unique snowflake (usually that's not a good thing). So we can't give you an exact step-by-step guide of how to daggerize your project. But we can share the collective experience of the Dagger community, distilled into a simple framework.
 
-Not sure which part of your pipeline to Daggerize first? Reach out to us on [Discord](https://discord.gg/dagger-io) via our [help forum](https://discord.com/channels/707636530424053791/1030538312508776540). Our team and your fellow Daggernauts will be happy to provide more specific guidance on your use case there! 
+## Phase 1: research
```
Comment (levlaz) [inline docs/current_docs/adopting.mdx:11]: Right before this section might be a good place to put a simple diagram showing the three stages.

File: docs/current_docs/adopting.mdx [hunk 2]
```diff
@@ -4,13 +4,121 @@ slug: /adopting
...
 
-Want to see how others have used Dagger? Check out our [case studies](https://dagger.io/resources/Case-Studies) or watch [demos](https://dagger.io/resources/Videos) from community members like yourself. 
+### Join our Discord!
 
-Do you have a use case that you’d like to talk about privately? Reach out to us at solutions@dagger.io
+Before anything, drop what you're doing and join our [Discord server](https://discord.gg/dagger-io). It is the beating heart of the Dagger community, and one of the killer features of the platform. It is the best place to ask questions, get inspiration, and ask for help when you are stuck.
+
+Consider yourself warned: adopting Dagger without joining our Discord is certainly possible, but you probably won't be as productive, and it definitely won't be as fun!
+
+### Look for inspiration
```
Comment (levlaz) [inline docs/current_docs/adopting.mdx:19]: We should curate this a bit more, there are many case studies and videos that predate modules, so folks finding those will lead to confusion. 

Perhaps a good place to highlight some of the mature DtW projects? 

If we do link out to more generic indexes, we should provide a pointer letting people know to limit their exploration to things that were published recently.

File: docs/current_docs/adopting.mdx [hunk 3]
```diff
@@ -4,13 +4,121 @@ slug: /adopting
...
+Before going any further, you should look for reasons *not* to adopt Dagger.
+
+Your project may not be a good fit for Dagger if:
+
+- It is so early that it doesn't need CI, and nobody worries about pipelines yet
+- Your *only* output is a Windows, Mac, iOS or Android application.
+- You are happily using a monolithic toolchain, such as Gradle, Nix or Bazel, with no exception and no fragmentation within the team
+- Your pipelines are heavily dependent on Windows or Mac runners
+
+In doubt, don't hesitate to ask the community - you've joined the Discord, right?
```
Comment (levlaz) [inline docs/current_docs/adopting.mdx:36]: I love the conversational nature of this :)

File: docs/current_docs/adopting.mdx [hunk 4]
```diff
@@ -4,13 +4,121 @@ slug: /adopting
...
+1. It suffers from one of these hair-on-fire problems:
+
+- It can't run across dev and CI
+- It can't run across different CI
+- It awkwardly glues together multiple systems, which makes it brittle
+- Not enough people know how it works
+- It's very slow and inefficient because of poor caching
+- It's so painful to change, that nobody wants to change it
+
+2. It can be Daggerized within a week
```
Comment (levlaz) [inline docs/current_docs/adopting.mdx:57]: There is some inconsistent casing for Daggerize vs daggerize, can we pick one?

File: docs/current_docs/adopting.mdx [hunk 5]
```diff
@@ -4,13 +4,121 @@ slug: /adopting
...
+
+Depending on the size and constraints of your project, this first pipeline could perform a build, test, deployment, or all the above. It could also automate a custom project-specific task.
+
+### Choosing a language
+
+When choosing the language to use for your pipeline, follow these principles:
+
+- *Optimize for participation*. The more people on the team are likely to participate in the development of the pipeline, the better.
+
+- *Check SDK availability*. Dagger supports 3 stable SDKs (Python, Go, Typescript), but it also has several community-contributed SDKs, some of which are actively used in production despite not being officially supported. Take the time to investigate the status of available SDKs, and ask the community for their opinion. Our Discord server is a good place to start - each supportd language has a dedicated channel.
```
Comment (levlaz) [inline docs/current_docs/adopting.mdx:69]: typo "supported"

```suggestion
- *Check SDK availability*. Dagger supports 3 stable SDKs (Python, Go, Typescript), but it also has several community-contributed SDKs, some of which are actively used in production despite not being officially supported. Take the time to investigate the status of available SDKs, and ask the community for their opinion. Our Discord server is a good place to start - each supported language has a dedicated channel.
```

File: docs/current_docs/adopting.mdx [hunk 6]
```diff
@@ -4,13 +4,121 @@ slug: /adopting
...
+
+### Integrating with your dev toolchain
+
+Once your pipeline is daggerized, in the form of one or more Dagger modules, it's time to integrate it with your development toolchain. Dagger is designed for flexibility, so you have many choices in how to integrate it. The downside is that it may feel like *too much* flexibility. What is the *correct* way to integrate?
+
+Here are a few guidelines:
+
+- If your team is willing and able to use the `dagger` CLI directly, we recommend that. Typically that means documenting a few `dagger call` commands in your README, explaining how to use `dagger functions` to discover more functions, and linking to the Dagger documentation to learn more.
+
+- If your team prefers (or is forced to) use another tool, whether off-the-shel like `make`, or custom-built tool or script, it's perfectly fine to wrap your `dagger call` invocations in existing Makefile, configuration files or scripts. Some power users may scratch beneath the surface and call `dagger call` directly; others won't. Both are fine.
```
Comment (levlaz) [inline docs/current_docs/adopting.mdx:91]:
```suggestion
- If your team prefers (or is forced to) use another tool, whether off-the-shelf like `make`, or custom-built tool or script, it's perfectly fine to wrap your `dagger call` invocations in existing Makefile, configuration files or scripts. Some power users may scratch beneath the surface and call `dagger call` directly; others won't. Both are fine.
```

File: docs/current_docs/adopting.mdx [hunk 7]
```diff
@@ -4,13 +4,121 @@ slug: /adopting
...
+Since your daggerized pipelines are packaged as Dagger Modules, and modules only have one dependency - the `dagger` CLI - integrating your daggerized pipelines boils down to three steps:
+
+1. Decide which CI event should trigger which Dagger pipeline
+2. For each pipeline to trigger, determine how to map inputs from the CI environment into arguments to the Dagger functions composing the pipeline
+3. Write the resulting `dagger call` command for each pipeline.
+4. Find the the [integration page](integrations) for your CI platform, and follow the instructions. For example: [GitHub Actions](integrations/github-actions.mdx), [GitLab CI](integrations/gitlab.mdx), [CircleCI](integrations/circleci.mdx), [Jenkins](integrations/jenkins.mdx), [Azure Pipelines](integrations/azure-pipelines.mdx), [AWS CodeBuild](integrations/aws-codebuild.mdx), [Argo Workflows](integrations/argo-workflows.mdx) and [Tekton](integrations/tekton.mdx).
+
+### The Demo
+
+Once your POC is successful, it's time for a victory lap! Make sure to show your team what you accomplished, why, and how. If you'd like help to make your presentation shine, don't hesitate to ask on Discord!
```
Comment (levlaz) [inline docs/current_docs/adopting.mdx:108]: nit: I think there are too many exclamation points in this section, I'd remove both of the ones here and leave the one at the end of the next paragraph.

File: docs/current_docs/adopting.mdx [hunk 8]
```diff
@@ -4,13 +4,121 @@ slug: /adopting
...
+
+If you'd like to, and are allowed, you can also show a demo at our bi-weekly community calls. If there's one thing Daggernauts love, it's a demo of a real world use of Dagger!
+
+## Phase 3: expand
+
+Once your POC is complete, it's time to expand your adoption of Dagger! Depending on your situation, there are 2 strategies you can adopt: incremental, or all-in.
+
+### Incremental expansion
+
+This is the safest, and most common strategy. Once you've successfully daggerized a small pipeline with your POC... Just repeat the process with another piece; then another one; etc. Depending on the constraints of your project, the genuine need for Dagger, and your team's priorities, this process could be spread out over months or even years; or it could snowball and spread like wildfire. We've seen both extremes, and will happily support you regardless of how fast you expand.
```
Comment (levlaz) [inline docs/current_docs/adopting.mdx:118]: nit: mixed metaphors :)

File: docs/current_docs/adopting.mdx [hunk 9]
```diff
@@ -4,13 +4,121 @@ slug: /adopting
...
+3. Write the resulting `dagger call` command for each pipeline.
+4. Find the the [integration page](integrations) for your CI platform, and follow the instructions. For example: [GitHub Actions](integrations/github-actions.mdx), [GitLab CI](integrations/gitlab.mdx), [CircleCI](integrations/circleci.mdx), [Jenkins](integrations/jenkins.mdx), [Azure Pipelines](integrations/azure-pipelines.mdx), [AWS CodeBuild](integrations/aws-codebuild.mdx), [Argo Workflows](integrations/argo-workflows.mdx) and [Tekton](integrations/tekton.mdx).
+
+### The Demo
+
+Once your POC is successful, it's time for a victory lap! Make sure to show your team what you accomplished, why, and how. If you'd like help to make your presentation shine, don't hesitate to ask on Discord!
+
+If you'd like to, and are allowed, you can also show a demo at our bi-weekly community calls. If there's one thing Daggernauts love, it's a demo of a real world use of Dagger!
+
+## Phase 3: expand
```
Comment (levlaz) [inline docs/current_docs/adopting.mdx:112]: I think this is a good place to encourage people to make reusable modules. Ideally abstract public ones that contribute to the daggerverse, but even private ones for internal use should be encouraged at this phase.

**Review by levlaz — DISMISSED** [review]
This is excellent overall, I love the conversational nature and the clear direction that it provides to teams of all sizes. 

I left a few comments and typo fixes.

File: docs/current_docs/adopting.mdx (hunk 1 above)
Comment (shykes) [reply docs/current_docs/adopting.mdx:11]: Good idea. Let's leave it for a follow-up PR if you don't mind?

File: docs/current_docs/adopting.mdx (hunk 2 above)
Comment (shykes) [reply docs/current_docs/adopting.mdx:19]: Makes sense. I'll also leave that for a follow-up PR (trying to get this one merged before 0.12, so we benefit from the influx of new visitors drawn by the announcement)

File: docs/current_docs/adopting.mdx (hunk 9 above)
Comment (shykes) [reply docs/current_docs/adopting.mdx:124]: Great idea. Done.

File: docs/current_docs/adopting.mdx (hunk 8 above)
Comment (shykes) [reply docs/current_docs/adopting.mdx]: You're right! Fixed.

**Comment by shykes** [comment]
- Incorporated @levlaz feedback, thank you!
- Added a phase 4: spin out module
- Added a ton of details about CI integration. Was shocked to find that some of this stuff is not written down anywhere. A good reminder that we still have massive gaps.

**Comment by shykes** [comment]
Failed check is a flake

File: docs/current_docs/adopting.mdx [hunk 10]
```diff
@@ -4,13 +4,147 @@ slug: /adopting
...
+- It awkwardly glues together multiple systems, which makes it brittle
+- Not enough people know how it works
+- It's very slow and inefficient because of poor caching
+- It's so painful to change, that nobody wants to change it
+
+2. It can be daggerized within a week
+
+3. You have the authority to daggerize it
+
+4.
```
Comment (jedevc) [inline docs/current_docs/adopting.mdx:57]: :eyes:

**Comment by shykes** [comment]
I'm going to fast track this since it got several thumbs up, and I'd like to take it live before 0.12 is announced

File: docs/current_docs/adopting.mdx (hunk 10 above)
Comment (shykes) [reply docs/current_docs/adopting.mdx:57]: gah

Comment (shykes) [reply docs/current_docs/adopting.mdx]: fixed, thank you

**Review by jedevc — APPROVED** [review]
One tiny typo nit, otherwise LGTM, so expediting :tada:

**Comment by shykes** [comment]
Thank you @jedevc ! Fixed the typo you found, rebased, and now merging!

**Pull request merged**